Add Typescript types to codebase

ninjha01 opened the pull request.

Commit:
[wip] adding type annotations

Commit:
seqviz.tsx typechecks

Commit:
translation.tsx typechecks

Commit:
Circular.tsx typechecks

Commit:
SeqViewer typechecks

ninjha01 committed:
Selection typechecks

ninjha01 committed:
[wip] seqblock

ninjha01 committed:
Create main.yml

ninjha01 committed:
use eslint

ninjha01 committed:
Merge branch 'cicd' into typescript

ninjha01 committed:
run cicd

ninjha01 committed:
run tsc in cicd

ninjha01 committed:
Add typescript types and run prettier

[wip] primers

[wip] index

[wip] zoom

[wip] infinite scroll

[wip] some util .ts files

run prettier on all files

temp

ninjha01 commented:
Apologies for the big PR - this adds typescript annotations and a CICD pipeline that includes formatting/linting/type checks/runs tests.

jjti commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,34 @@
+    #   run: npm run build
+    # deploy:
+    # - name: Deploy to Staging
+    #   run: ???
```
this will be cool. Can do the npm deploy here

jjti commented on .prettierrc:
```diff
@@ -0,0 +1,6 @@
+  "tabWidth": 4,
+  "semi": false,
+  "singleQuote": true
+}
```
there's already a prettier definition in `package.json`: https://github.com/Lattice-Automation/seqviz/blob/develop/package.json#L52-L55

It differs slightly. I'd prefer that we keep the existing formatting. If you want to change anything that's fine, but lets change it there and do a change just for the auto-formatting (so this commit is focused on the TS changes)

ninjha01 replied:
This makes sense to me - unfortunately I applied formatting changes throughout the process of writing this PR (I run it on save). If it's an issue, I can try seeing if knocking prettier down to 1.19.1 and running it using the old config will cut down on the diff size.

jjti replied:
I think the CI stuff and prettier formats are cool, but also separate from the goal of adding TypeScript. They should both be their own PRs/changes

Re: prettier in particular, I'd try to change the formatting as little as possible. If bumping prettier changes formatting (?) with existing config that's one thing, but changing defaults is unnecessary

ninjha01 replied:
I'll make a separate PR for the CI, and commit 36bbab4 does what I outlined above, but doesn't reduce the diff dramatically. Happy to try another approach.

jjti replied:
Afaik decrementing prettier to what it had been before and removing the .prettier.rc should revert formatting to the way it had been. Did anything change in prettier 2.0 other than defaults?

Do you need to also remove the pre-commit hook?

jjti replied:
Or we could do the prettier bump first in a separate CR first, this all second. That's probably easier

ninjha01 replied:
Great idea - I'll pull the pre-commit hook and run prettier in a separate PR

ninjha01 replied:
@JJTimmons just merged `develop` into this branch and the diff is more reasonable in size. (turned out the bulk of the changed lines came from running prettier on a big example json file)

jjti reviewed (changes requested):
From what I can tell things look good, but the diff in prettier configs means that most of the CR is auto-formatting diffs. Can you please checkout that prettier comment to remove the formatting changes?

jjti commented on src/SeqViz/Circular/AnnotationsViewer.tsx:
```diff
@@ -2,10 +2,38 @@ import * as React from "react";
+import { inputRefFuncType } from "../Linear/SeqBlock/Translations";
+import { Annotation, Label } from "../../part";
+
+interface AnnotationsViewerProps {
+  radius: number;
+  center: Coor;
+  lineHeight: number;
+  seqLength: number;
+  findCoor: (index: number, radius: number, rotate?: boolean) => Coor;
+  getRotation: (index: number) => string;
+  generateArc: (args: {
+    innerRadius: number;
+    outerRadius: number;
+    length: number;
+    largeArc: boolean; // see svg.arc large-arc-flag
+    sweepFWD?: boolean;
+    arrowFWD?: boolean;
+    arrowREV?: boolean;
+    offset?: number;
+  }) => string;
+  rotateCoor: (coor: Coor, degrees: number) => Coor;
+  inputRef: inputRefFuncType;
+  annotations: Annotation[];
+  size: SizeType;
+  rowsToSkip: number;
+  inlinedAnnotations: Label[];
+}
 
 /**
  * Used to build up all the path elements. Does not include a display
- * of the annotation name or a line connecting name to annotation
+ * of the annotation name or a line connecting name to annotatino
```
nit (honestly you can ignore anything I post with nit, not worth a rev by itself, just tagging stuff in case it slid by)

jjti commented on src/SeqViz/Circular/AnnotationsViewer.tsx:
```diff
@@ -14,9 +42,9 @@ import { COLOR_BORDER_MAP, darkerColor } from "../../utils/colors";
  *
  * @type {Function}
  */
-export default class Annotations extends React.PureComponent {
+export default class AnnotationsViewer extends React.PureComponent<AnnotationsViewerProps> {
```
I think I get the intent of this: `Annotations` like the data format/interface, versus the thing that renders them in the Linear/Circular viewers. But if keeping those separate I think I'd name the data format/interface as `IAnnotation`, `IIndex`, etc

The React component names make sense in the context of where they show up within the viewer. Like if we were to make a change later on to let users pass in the child components of the `Linear/Circular` components, I think that the way things are now, below, is more intuitive than it would be after the change (down below)

```
<LinearViewer>
  <Index />
  <Annotations />
</LinearViewer>
```

vs

```
<LinearViewer>
  <Index />
  <AnnotationsViewer />
</LinearViewer>
```

jjti commented on src/SeqViz/Linear/InfiniteScroll.tsx:
```diff
@@ -2,31 +2,42 @@ import * as React from "react";
+  totalHeight: number;
+  size: SizeType;
+  bpsPerBlock: number;
+}
+interface InfiniteScrollState {
```
nit for space between interfaces. And interfaces and comments (down below)

jjti commented on src/SeqViz/Linear/SeqBlock/Translations.tsx:
```diff
@@ -3,44 +3,63 @@ import * as React from "react";
+  direction: -1 | 1;
+}
+
+export type inputRefFuncType = <T>(id: string, ref: unknown) => React.LegacyRef<T>;
+
+export type findXandWidthType = (
+  n1: number | undefined,
+  n2?: number
+) => {
+  x: number;
+  width: number;
+};
```
These should be exported from where that method is defined in `SeqBlock`. Or, if there's a dependency cycle there, from a types-only file in the `SeqBlock` dir

jjti commented on src/SeqViz/Linear/SeqBlock/Translations.tsx:
```diff
@@ -3,44 +3,63 @@ import * as React from "react";
+
+export type inputRefFuncType = <T>(id: string, ref: unknown) => React.LegacyRef<T>;
+
+export type findXandWidthType = (
```
I've only seen types capitalized with upper camel case

jjti commented on src/SeqViz/Linear/SeqBlock/Translations.tsx:
```diff
@@ -70,31 +89,46 @@ export default class TranslationRows extends React.PureComponent {
+  y: number;
+  height: number;
+  id?: string;
+}
+class TranslationRow extends React.Component<TranslationRowProps> {
```
nit on the space

jjti reviewed (changes requested):
Some minor comments from after scrolling thru

ninjha01 committed:
code review suggestions

[wip] rename annotations/index/find components back

[wip] inputreffunctype and findxandwidthtype changes

ninjha01 commented:
Thanks for the prompt review @JJTimmons ! I believe I've implemented the  requested changes in the latest commit (aside from the prettier-related one). Let me know if there's any other changes you'd like .

ninjha01 committed:
pull out cicd

ninjha01 committed:
drop down to prettier 1.19.1

ninjha01 committed:
run prettier@2.5.1 on all files

ninjha01 committed:
remove pre-commit-config.yaml

ninjha01 committed:
pre-commit hook

ninjha01 committed:
Merge branch 'prettier' into typescript

ninjha01 committed:
run prettier on source of big diff in typescript branch

ninjha01 committed:
Merge branch 'prettier' into typescript

ninjha01 committed:
Merge branch 'develop' into typescript

ninjha01 commented:
Hey @JJTimmons, just wanted to double check that things looked good from your end.

jjti commented:
I'm still seeing 99% formatting changes when I checkout the PR's changes: https://github.com/Lattice-Automation/seqviz/pull/122/files

I see that we merged in changes from those branches, wonder if we don't need to rebase from `develop` again though. I haven't used Github day to day in a couple years though. Let me know if I'm seeing outdated changes

ninjha01 commented:
> I see that we merged in changes from those branches, wonder if we don't need to rebase from `develop` again though. I haven't used Github day to day in a couple years though. Let me know if I'm seeing outdated changes

No, you're right. Just rebased and the diff looks cleaner

jjti commented on src/SeqViz/Linear/SeqBlock/SeqBlock.tsx:
```diff
@@ -1,12 +1,65 @@
 import * as React from "react";
-
-import Annotations from "./Annotations";
+import { MouseEventHandler } from "react";
```
why not use `React.MouseEventHandler` since we're already importing (above)

jjti commented on src/SeqViz/Circular/Annotations.tsx:
```diff
@@ -1,7 +1,34 @@
+import { InputRefFuncType } from "../Linear/SeqBlock/SeqBlock";
+import { Coor, SizeType } from "./Circular";
+
+interface AnnotationsViewerProps {
```
should be `AnnotationsProps`

jjti commented on src/SeqViz/Circular/Annotations.tsx:
```diff
@@ -14,9 +41,9 @@ import { COLOR_BORDER_MAP, darkerColor } from "../../utils/colors";
  *
  * @type {Function}
```
you've done a good job of this already, but would remove whatever shoddy jsdocs you see lying around. Was from a time of Flow (and even slightly before then)

jjti commented on src/SeqViz/Circular/CircularSelection.tsx:
```diff
@@ -1,6 +1,33 @@
+import { string } from "prop-types";
```
what is this? Isn't string a normal ts type?

jjti commented on src/SeqViz/Circular/CutSites.tsx:
```diff
@@ -1,8 +1,31 @@
 import * as React from "react";
+import { InputRefFuncType } from "../Linear/SeqBlock/SeqBlock";
+import { Coor, ICutSite } from "./Circular";
+interface CutSitesProps {
```
nit: missing space

jjti commented on src/SeqViz/Circular/CutSites.tsx:
```diff
@@ -1,8 +1,31 @@
 import * as React from "react";
+import { InputRefFuncType } from "../Linear/SeqBlock/SeqBlock";
```
nit on space between external and internal imports. May as well go one way or the other across all files, and so far it's w/ a newline gap/space

jjti commented on src/SeqViz/Circular/Find.tsx:
```diff
@@ -1,13 +1,38 @@
+import { Coor } from "./Circular";
 
-export default class CircularFind extends React.PureComponent {
+interface CircularFindProps {
```
why not just `FindProps` since it's not exported, seen elsewhere?

jjti commented on src/SeqViz/Linear/Linear.tsx:
```diff
@@ -115,7 +141,7 @@ class Linear extends React.Component {
      * @param {*} annotations
      * @return annotations
      */
```
nit: more jsdocs to remove. nbd but will be nice to be fully ts w/o these floating around

jjti commented on src/SeqViz/Linear/SeqBlock/Selection.tsx:
```diff
@@ -180,22 +197,23 @@ export class Block extends React.PureComponent {
+      return null;
+    } else {
+      return (
+        <React.Fragment>
```
nit: I think we can roll w/ the shorter syntax now: https://reactjs.org/docs/fragments.html#short-syntax

jjti commented on src/SeqViz/Linear/SeqBlock/Selection.tsx:
```diff
@@ -180,22 +197,23 @@ export class Block extends React.PureComponent {
-        {secondBlock}
-      </React.Fragment>
-    );
+    if (!x || !width) {
```
this is diff from behavior above. Just checking but you tested the viewers w/ this change right? (`npm run` + messed around w/ viewers)

jjti commented on src/SeqViz/Linear/SeqBlock/Selection.tsx:
```diff
@@ -180,22 +197,23 @@ export class Block extends React.PureComponent {
   }
 }
-
 export default { Edges, Block };
```
nit: leave the space

jjti commented on src/SeqViz/Linear/SeqBlock/SeqBlock.tsx:
```diff
@@ -1,12 +1,65 @@
+  width: number;
+};
+
+interface SeqBlockPosition {
+  x: number; // [the x positioning, from left...]
```
nit: if you comment above the var or with `/* */` I think VSCode will pick it up and add to the hover-over tips. I don't _think_ that the same happens for inline comments like this (could be wrong)

jjti commented on src/SeqViz/Linear/SeqBlock/SeqBlock.tsx:
```diff
@@ -1,12 +1,65 @@
+  charWidth: number; // [the width of each character in the element]
+}
+
+interface SeqBlockProps {
```
I apologize about the size of this beast. Having the TS will make the changes much safer tho

jjti commented on src/SeqViz/Linear/SeqBlock/SeqBlock.tsx:
```diff
@@ -21,41 +74,33 @@ import TranslationRows from "./Translations";
   };
 
   /**
-   * @typedef {Object} SeqBlockPosition
+   * @typedef {Object}
    * @property {Number}     x              [the x positioning, from left...]
    * @property {Number}     width          [the width of the passed element]
    * @property {Number}     charWidth      [the width of each character in the element]
    */
```
nit: more jsdoc fluff to nuke

jjti commented on src/SeqViz/Linear/SeqBlock/SeqBlock.tsx:
```diff
@@ -21,41 +74,33 @@ import TranslationRows from "./Translations";
+ *
+ * a helper method that's used in several of the child components to figure
+ * out how far from the left the element is and how wide it should be
+
```
nit

jjti commented on src/SeqViz/SeqViewer.tsx:
```diff
@@ -28,17 +45,16 @@ See: https://github.com/Lattice-Automation/seqviz#optionsstyle-`);
 
   /** this is here because the size listener is returning a new "size" prop every time */
-  shouldComponentUpdate = (nextProps, nextState) => !isEqual(nextProps, this.props) || !isEqual(nextState, this.state);
+  shouldComponentUpdate = (nextProps: SeqViewerProps, nextState: unknown) =>
```
shouldn't we type out the state here as well?

https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/class_components

jjti commented on src/SeqViz/SeqViewer.tsx:
```diff
@@ -76,7 +92,7 @@ See: https://github.com/Lattice-Automation/seqviz#optionsstyle-`);
       size,
       zoom: { linear: zoom },
-      Linear: true,
+      Linear: true
```
prettier?

jjti commented on src/SeqViz/SeqViz.tsx:
```diff
@@ -179,7 +180,7 @@ export default class SeqViz extends React.Component<SeqVizProps, any> {
   /**
    * Search for the query sequence in the part sequence, set in state
    */
-  search = (part: Part | null = null) => {
+  _search = (part: Part | null = null) => {
```
what is this for? Looks like python w/ the underscore to indicate a private method. Re: google style guide link below lets avoid the underscore prefix

ninjha01 replied:
This was an attempt to distinguish from the local `search` var created inside the search method - it was a little confusing keep the method, the state var, and the props var straight. I'll rename it back unless you have an alternate name you prefer.

jjti commented on src/SeqViz/SeqViz.tsx:
```diff
@@ -265,10 +263,9 @@ export default class SeqViz extends React.Component<SeqVizProps, any> {
-    // @ts-expect-error ts-migrate(2345) FIXME: Argument of type 'string | undefined' is not assig... Remove this comment to see the full error message
-    if (getSeqType(seq) === "dna") {
-      compSeq = compSeq || part.compSeq || dnaComplement(seq).compSeq;
+    const _seq: string = seq || part.seq || "";
```
the `_prefix` identifiers are not normal in TS, recommended against. Let's remove or find a diff way to avoid duplicate identifier names https://google.github.io/styleguide/tsguide.html#identifiers

jjti commented on src/io/parsers/genbank.ts:
```diff
@@ -134,7 +134,6 @@ export default async (fileInput, fileName, colors = []) =>
               annotations.push({
-                // @ts-expect-error ts-migrate(2345) FIXME: Argument of type 'never[]' is not assignable to pa... Remove this comment to see the full error message
                 ...annotationFactory(row_i, colors),
                 // @ts-expect-error ts-migrate(2322) FIXME: Type 'any' is not assignable to type 'never'.
```
is this one still needed?

jjti commented on src/utils/colors.ts:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck
```
why not check this file? Is used a lot

ninjha01 replied:
Honestly - the `pSBC` function grabbed from https://github.com/PimpTrizkit/PJs/wiki/12.-Shade,-Blend-and-Convert-a-Web-Color-(pSBC.js) looked really unpleasant to annotate, and typescript can't ignore blocks of code last i checked, just entire files.

jjti replied:
ack and understood. I'd consider moving to its own file since we use the other functions in a few places. Agreed w/ pSBC being a bear: I copied it uncritically from SO

jjti commented on src/utils/parser.ts:
```diff
@@ -44,20 +46,11 @@ const DNAComplement = {
   u: "a",
 };
 
-/**
```
nice

jjti reviewed (changes requested):
One additional thing besides that changes below is that what consumers will care about is not these internal lib TS changes (that matter a lot for us as devs), but about the presence of a declaration file so they see the viewer props when they're developing against `SeqViz` within their own IDE.

I checked out the code at current state and found that the minimal set of changes needed for that are to:
1. uncomment the `declarations` setting in tsconfig (so: yes to declaration files)
2. point the `types` prop in package.json at ./dist/viewer.ts.d

https://www.typescriptlang.org/tsconfig#declaration

Can test the above by building the changes. Then cd'ing to the demo dir and installing like `npm install ..` and `mv App.jsx App.tsx`

ninjha01 committed:
code review requests

ninjha01 committed:
Merge branch 'develop' into restart-typescript-branch

ninjha01 committed:
export type declarations

ninjha01 committed:
debugging demo App.tsx

ninjha01 committed:
refactor pSBC func to seperate file

jjti commented on demo/package.json:
```diff
@@ -9,7 +9,7 @@
     "react-scripts": "^3.3.0",
     "semantic-ui-react": "^0.87.3",
-    "seqviz": "^3.0.25"
+    "seqviz": "file:.."
```
this is cool

jjti commented on src/SeqViz/Circular/Annotations.tsx:
```diff
@@ -1,7 +1,34 @@
+import { Annotation, Label } from "../../part";
 import { COLOR_BORDER_MAP, darkerColor } from "../../utils/colors";
+import CentralIndexContext from "../handlers/centralIndex";
+import { InputRefFuncType } from "../Linear/SeqBlock/SeqBlock";
```
nothing in `/Circular` should import anything from `/Linear` or vice versa. For the types shared between them, define them higher up in the tree in a shared parent.. or a separate types file

jjti commented on src/SeqViz/Circular/Annotations.tsx:
```diff
@@ -57,31 +82,29 @@ export default class Annotations extends React.PureComponent {
-                ))
+              return (
+                <SingleAnnotation
+                  {...this.props}
```
nbd but I like how in other places you got rid of this destructuring. Doesn't have to change but I think it's cleaner the way you've been doing it, this is sort of tough to keep track of and makes debugging the TS errors tougher

jjti commented on src/SeqViz/Circular/Circular.tsx:
```diff
@@ -1,25 +1,92 @@
 import { stackElements } from "../elementsToRows";
 import withViewerHOCs from "../handlers";
 import CentralIndexContext from "../handlers/centralIndex";
+import { InputRefFuncType } from "../Linear/SeqBlock/SeqBlock";
```
same comment as above re: Circular not importing from Linear

jjti commented on src/SeqViz/Circular/Circular.tsx:
```diff
@@ -1,25 +1,92 @@
+export interface Primer {
+  direction: 1 | -1;
+}
+export interface ICutSite {
```
nit: space

jjti replied:
both of these interfaces look non-Circular specific. Like they're in both viewers (Primers are deprecated, can be neglected, but cut sites are def in common)

For search the results of searching were put in the search util file. We could do the same here with cut sites re: the enzymes file (or use some other pattern where these types are in a shared types file, doesn't matter).

jjti commented on src/SeqViz/Circular/Circular.tsx:
```diff
@@ -1,25 +1,92 @@
+  centralIndex: number;
+  setCentralIndex: (update: number) => void;
+}
+interface CircularState {
```
nit: space

jjti commented on src/SeqViz/Circular/Circular.tsx:
```diff
@@ -1,25 +1,92 @@
+  recogStrand: unknown;
+}
+
+export interface SizeType {
```
same thing as above

jjti commented on src/SeqViz/Circular/Circular.tsx:
```diff
@@ -194,24 +247,25 @@ class Circular extends React.Component {
+    const { innerRadius, outerRadius, length, largeArc, sweepFWD, arrowFWD, arrowREV, offset } = args;
     const { radius } = this.props;
     const { seqLength, lineHeight } = this.state;
+    const _offset = offset === undefined ? 0 : offset;
```
see prefix comment

jjti commented on src/SeqViz/Circular/Index.tsx:
```diff
@@ -13,15 +42,14 @@ import CentralIndexContext from "../handlers/centralIndex";
  * are too many elements in the circular viewer and the name won't fit
  */
-export default class Index extends React.PureComponent {
+export default class IndexViewer extends React.PureComponent<IndexProps> {
```
should be `Index`

jjti commented on src/SeqViz/Circular/Labels.tsx:
```diff
@@ -1,8 +1,36 @@
+  rotateCoor: (coor: Coor, degrees: number) => Coor;
+  inputRef: InputRefFuncType;
+}
+interface LabelsState {
```
nit space

jjti commented on src/SeqViz/Circular/Labels.tsx:
```diff
@@ -320,3 +347,18 @@ export default class Labels extends React.Component {
+  textCoor: Coor;
+  textAnchor: unknown;
+}
+interface GroupedLabelsWithCoors {
```
nit on space between interfaces

I'm posting this comment repeatedly/too much. Made this CR to help find/resolve these automagically: https://github.com/Lattice-Automation/seqviz/pull/135

jjti commented on src/SeqViz/Linear/InfiniteScroll.tsx:
```diff
@@ -37,15 +47,13 @@ export default class InfiniteScroll extends React.PureComponent {
   };
 
-  componentDidUpdate = (prevProps, prevState, snapshot) => {
+  componentDidUpdate = (prevProps: InfiniteScrollProps, prevState: any, snapshot: any) => {
```
wouldn't `prevState` be `InfiniteScrollState`?

jjti commented on src/SeqViz/Linear/Linear.tsx:
```diff
@@ -1,13 +1,58 @@
-
+import { SearchResult } from "../../utils/search";
 import { createLinearTranslations } from "../../utils/sequence";
+import { Coor, ICutSite, Primer, SizeType } from "../Circular/Circular";
```
re: earlier comment, these are all candidates for being moved up in the tree

ninjha01 committed:
[wip] code review requests

ninjha01 committed:
Merge branch 'develop' into typescript

ninjha01 committed:
code review request stragglers

ninjha01 commented:
Hey Josh, I've implemented the changes you requested - I'd like to get this merged soon so I can get started on the Sequence Highlighting feature you called out. Does that sound reasonable?

jjti commented on src/utils/search.ts:
```diff
@@ -4,13 +4,16 @@ import { nucleotideWildCards, nucleotides, reverse, translateWildNucleotides } f
 /**
  * Search the seq in the forward and reverse complement strands.
  * Return all matched regions. Accounts for abiguous BP encodings and allows for mismatches
```
this javadoc is not on the `SearchResult` type not the method

jjti commented on src/utils/search.ts:
```diff
@@ -43,14 +45,9 @@ export default (query, mismatch, seq) => {
- * @returns {[SeqReturn]}
  */
-const search = (query, subject, mismatch, fwd) => {
+
```
space

jjti commented on src/part.ts:
```diff
@@ -4,7 +4,16 @@ export interface Element {
   direction: number;
 }
 
+export interface Label {
```
the labels can go in circular or in circular/labels. Are created from parts' annotations but not associated with the parts themselves

jjti commented on src/SeqViz/CommonTypes.ts:
```diff
@@ -0,0 +1,30 @@
+import * as React from "react";
+
+export type InputRefFuncType = <T>(id: string, ref: unknown) => React.LegacyRef<T>;
+
```
uppercase file names indicate the file exports a React component. This could be `shared.ts` or `common.ts` or something like that

jjti commented on src/SeqViz/CommonTypes.ts:
```diff
@@ -0,0 +1,30 @@
+
+export type InputRefFuncType = <T>(id: string, ref: unknown) => React.LegacyRef<T>;
+
+export interface SizeType {
```
bit of a pref for `ISize` since this is an interface not a type

jjti commented on src/SeqViz/SeqViz.tsx:
```diff
@@ -6,13 +6,28 @@ import { Annotation, Element, Part } from "../part";
 import { SelectionContext, defaultSelection } from "./handlers/selection";
 import "./style.css";
 
+export interface SeqVizSelection {
```
This makes a bit more sense coming from `./handlers/selection`. Analogous to SearchResults in search

jjti commented on src/SeqViz/Circular/Labels.tsx:
```diff
@@ -320,3 +349,20 @@ export default class Labels extends React.Component {
   }
 }
+
+interface LabelWithCoors {
+  label: Label;
+  lineCoor: Coor;
+  textCoor: Coor;
+  textAnchor: unknown;
+}
+
+interface GroupedLabelsWithCoors {
```
lets put at top since they're used thru rest of file. Or immediately before the first class to use them

jjti reviewed (changes requested):
Only a few smaller things left (to be expected for a +/- 7k CR). I think it's close. Also there are some unresolved comments from earlier reviews, please check those out

My other qq/hope is that you've tested all of this (eg via the demo site). *Or just use `npm run start` and modify that index.ts file's settings to try out both viewers.

jjti commented:
quick note/pointer about the merge conflicts

ninjha01 commented:
Hey @JJTimmons 

I've tested this branch by running `npm run start` on both the `develop` branch and the `typescript` branch. The only change I can see is this misplaced/mis-styled annotation in the `typescript` branch. Any idea what could have caused that?

`develop`: 
<img width="1253" alt="Develop Branch" src="https://user-images.githubusercontent.com/25281692/154506128-a1cba893-467f-413d-8cad-aec4c0457fa1.png">

`typescript`:
<img width="1252" alt="Typescript Branch" src="https://user-images.githubusercontent.com/25281692/154506184-552a2a7c-074a-4c04-9e54-3780d6acf4a7.png">


Additionally, (on `develop`) the demo fails to fetch data from iGem with the following error.

```
index.js:1 Error: Failed to retrieve a seq with accession BBa_K1598008 from http://parts.igem.org/xml/part.BBa_K1598008
    at index.js:6:1
    at index.js:6:1
    at Object.next (index.js:6:1)
at r (index.js:6:1)
```
I'll make a seperate issue to fix this, and can tackle that once this PR gets merged in.

I think I've also addressed the most recent round of changes you requested. Let me know what you think!

jjti commented:
Re: annotation in typescript branch I have no idea offhand, looks position or width related. Could be a missing prop. Worth debugging from within that class.

For testing I'd also turn on a couple translations to make sure those still look right, and do a find as well (ie hit all the child classes at least once)

Re: igem it could be the same as in this issue: https://github.com/Lattice-Automation/seqviz/issues/105

jjti commented on src/SeqViz/handlers/events.tsx:
```diff
@@ -76,7 +74,7 @@ const withEventRouter = WrappedComp =>
      * 	Up, Right, Down, Left: some directional movement of the cursor
      *
-     * @param {String} i  		  one of the commands listed above
+     
```
nbd and not worth a revision over, but we left the newline characters here and in some other places below. can regex replace all to make fix quick

ninjha01 commented:
> quick note/pointer about the merge conflicts

Are you seeing merge conflicts? I'm not seeing any on my end.

jjti commented:
I'm just pointing out the GitHub PR comment "This branch cannot be rebased due to conflicts". Again I haven't used GitHub day to day in years, and even then, often w/o PRs. But taken literally it looks like you need to rebase develop onto typescript locally first (then will see and need to resolve conflicts)

ninjha01 commented:
Curious - I don't see that comment: 
<img width="995" alt="image" src="https://user-images.githubusercontent.com/25281692/154534731-507b8b71-1053-41db-918e-21b0818cbfdd.png">

jjti commented:
huh yeah that's weird, agreed
<img width="952" alt="Screen Shot 2022-02-17 at 12 14 41 PM" src="https://user-images.githubusercontent.com/13923102/154534886-df5ceb5d-47ab-4851-8ece-34c7ca473b3c.png">

jjti commented:
Ah I have rebase as default for PRs, not squash

ninjha01 commented:
Was just about to comment the same thing - do you have a strong preference for one over the other?

jjti commented:
I don't, no

jjti commented:
I'm going AWOL but the PR looks good after once-over testing all the circular and linear functionality. Some of these lack enzyme tests so need to be tested manual (on account of frontend testing being a pita):
- search
- cut sizes
- annotations across multiple rows
- selections across seq blocks
- translations across seq blocks

ninjha01 commented:
Sweet, I'll make sure to go through those workflows.

Thanks for your time and review!

ninjha01 committed:
debugging missing annotation
